### PR TITLE
Implement new tests

### PR DESCRIPTION
**Pull request by bugagashenkj**
#110

### CODE REVIEW

File: test/database.js [hunk 1]
```diff
@@ -1,17 +1,80 @@
 'use strict';
 
+/* eslint new-cap: 0 */
+const metatests = require('metatests');
 const metaschema = require('..');
+const decorators = require('../lib/decorators');
```
Comment (tshemsedinov) [inline test/database.js]: `{ Include, ... } = require('../lib/decorators').attribute;`

File: test/database.js [hunk 2]
```diff
@@ -1,17 +1,80 @@
...
+  metaschema.load('general', (err, schema) => {
+    test.assertNot(err);
+    metaschema.build(schema);
+
+    for (const categoryName in categoriesConfigs) {
+      const config = categoriesConfigs[categoryName];
+      const category = metaschema.categories.get(categoryName);
+
+      test.strictSame(config.name, category.name);
+      for (const key in config.definition) {
```
Comment (lundibundi) [inline test/database.js]: Wouldn't `test.strictSame(category.definition, config.definition)` be enough?

File: test/database.js [hunk 3]
```diff
@@ -1,17 +1,80 @@
...
+metatests.test('Database / database', (test) => {
+  metaschema.load('general', (err, schema) => {
+    test.assertNot(err);
+    metaschema.build(schema);
+
+    for (const categoryName in categoriesConfigs) {
+      const config = categoriesConfigs[categoryName];
+      const category = metaschema.categories.get(categoryName);
+
+      test.strictSame(config.name, category.name);
```
Comment (lundibundi) [inline test/database.js]: Argument order is incorrect, its `actual, expected`. This should be `test.strictSame(category.name, config.name);`

File: test/database.js [hunk 4]
```diff
@@ -1,17 +1,80 @@
...
+    Languages: Many({
+      category: 'Language',
+      definition: categoriesConfigs.Language.definition
+    })
+  }
+};
+
+metatests.test('Database / database', (test) => {
+  metaschema.load('general', (err, schema) => {
+    test.assertNot(err);
```
Comment (lundibundi) [inline test/database.js]: test.error(err);

File: test/database.js [hunk 5]
```diff
@@ -1,17 +1,80 @@
...
+    Sex: domains.sex,
+    Born: domains.born,
+    Languages: Many({
+      category: 'Language',
+      definition: categoriesConfigs.Language.definition
+    })
+  }
+};
+
+metatests.test('Database / database', (test) => {
```
Comment (lundibundi) [inline test/database.js]: Test case name doesn't seem correct.

File: test/guard.js [hunk 6]
```diff
@@ -15,9 +16,9 @@ const buyTickets = (
   return true;
 };
 
-buyTickets(101, 2, 'Kiev, Pobedy 37');
-try {
-  buyTickets(101, '2', 37);
-} catch (e) {
-  console.log(e.message);
-}
+metatests.case('Metaschema / guard', { buyTickets }, {
```
Comment (belochub) [inline test/guard.js]: This doesn't really test anything since `guard` is always returning `true`.

File: test/structure.js [hunk 7]
```diff
@@ -1,61 +1,56 @@
 'use strict';
 
+const metatests = require('metatests');
 const metaschema = require('..');
 
 metaschema.load('geometry', (err, schema) => {
   if (err) throw err;
```
Comment (belochub) [inline test/structure.js]: This should be tested as well.

**Comment by lundibundi** [comment]
@bugagashenkj could you rebase this and update to use metatests@0.2.1?

File: test/structure.js [hunk 8]
```diff
@@ -1,61 +1,59 @@
 'use strict';
 
+const metatests = require('metatests');
 const metaschema = require('..');
 
 metaschema.load('geometry', (err, schema) => {
```
Comment (lundibundi) [inline test/structure.js]: `metatests` don't support such usage and this may lead to multiple 'results' printed in the end.
All tests should be added synchronously. Therefore this must be changed to smth like
```js
const geometryTest = metatests.test('geometry');
metaschema.load('geometry', (err, schema) => {
  // ... use geometryTest and/or metaschema.case ...
  // ...
  geometryTest.end();
})
```

Or
```js
metatests.test('geometry', (test) => {
  metaschema.load('geometry', (err, schema) => {
    // ... use test and/or metaschema.case ...
    // ...
    geometryTest.end();
  })
});
```

Whichever you prefer.

File: test/structure.js [hunk 9]
```diff
@@ -1,61 +1,59 @@
...
-  const pl1 = geometry('Polyline', {
-    points: [
-      { x: 1, y: 3 },
-      { x: 2, y: 2 },
-      { x: 3, y: 4 },
-      { x: 4, y: 5 },
-      { x: 3, y: 5 }
+  const geometry = metaschema.build(schema);
+  metatests.case('Metaschema / structure', { metaschema, geometry }, {
+    'metaschema.createInstance': [
```
Comment (lundibundi) [inline test/structure.js]: Could probably be inlined to `createInstance` to be brief. (with appropriate change to the namespace argument)

File: test/structure.js [hunk 10]
```diff
@@ -1,61 +1,59 @@
...
+      ['Point', { x: 3, y: 5 }, { x: 3, y: 5 }]
+    ],
+    'geometry': [
+      ['Point', { x: 3, y: 5 },        { x: 3, y: 5 }],
+      ['Point', [3, 5],                { x: 3, y: 5 }],
+      ['Point', 3, 5,                  { x: 3, y: 5 }],
+      ['Point', { x: 'hello', y: 2 },  null],
+      ['Point', { name: 'Marcus' },    {}],
```
Comment (lundibundi) [inline test/structure.js]: Shouldn't this be `null`?
Looks like a bug as those fields are `required`.

File: test/validate.js [hunk 11]
```diff
@@ -1,9 +1,12 @@
 'use strict';
 
+const metatests = require('metatests');
 const metaschema = require('..');
 
 const path = './schemas/metaschema/StructureField.schema';
 
+const getDefinition = name => metaschema.categories.get(name).definition;
+
 metaschema.loadSchema(path, (err, schema) => {
```
Comment (lundibundi) [inline test/validate.js]: Same as in `test/structure.js`.

File: test/validate.js [hunk 12]
```diff
@@ -14,46 +17,33 @@ metaschema.loadSchema(path, (err, schema) => {
...
-  // Validate category field schema
-  {
-    const { definition } = metaschema.categories.get('StructureField');
-    const valid = metaschema.validateFields('StructureField', definition);
-    console.dir(valid);
-  }
-
+  metatests.case('Metaschema / validate', { metaschema }, {
+    'metaschema.validate': [
+      [
```
Comment (lundibundi) [inline test/validate.js]: Could you also add
`'Schema1', {}, { valid: true, errors: [] }`
and
`'Schema2', {}, { valid: false, errors .... }`
?

**Comment by lundibundi** [comment]
ping @bugagashenkj

File: test/structure.js (hunk 10 above)
Comment (bugagashenkj) [reply test/structure.js]: Should I do something about this?

**Comment by bugagashenkj** [comment]
@lundibundi please approve.

File: package.json [hunk 13]
```diff
@@ -23,7 +23,7 @@
   },
   "readmeFilename": "README.md",
   "scripts": {
-    "test": "npm run lint && npm run docs && node ./test/all.js",
+    "test": "eslint . && metatests test/",
```
Comment (lundibundi) [inline package.json]: Why did you remove the previous changes? This should be
`"test": "npm run lint && npm run docs && metatests test/"`.

File: test/structure.js (hunk 10 above)
Comment (lundibundi) [reply test/structure.js]: You can go ahead and change this as that PR is going to be accepted, there is only a minor change waiting.

**Comment by lundibundi** [comment]
Landed in https://github.com/metarhia/metaschema/pull/141 (https://github.com/metarhia/metaschema/pull/141/commits/bbe16021cd339eba7e33e189d7a07047c5767b30).
